graylog2-plugin-interfaces/src/main/java/org/graylog2/plugin/inputs/transports/AbstractTcpTransport.java

```diff
@@ -149,12 +187,118 @@ protected Bootstrap getBootstrap() {
                 } catch (SSLException e) {
                     LOG.error("Error creating SSL context. Make sure the certificate and key are in the correct format: cert=X.509 key=PKCS#8");
                     throw e;
                 }
             }
+
+            private SSLEngine createSslEngine() throws FileNotFoundException, IOException, GeneralSecurityException {
+                SSLContext instance = SSLContext.getInstance("TLS");
+                TrustManager[] initTrustStore;
+                if ((tlsWantClientAuth || tlsNeedClientAuth) && tlsClientAuthCertFile.exists()) {
+                    initTrustStore = initTrustStore();
+                }
+                else {
+                    initTrustStore = new TrustManager[0];
+                }
+                instance.init(initKeyStore(), initTrustStore, new SecureRandom());
+                SSLEngine engine = instance.createSSLEngine();
+                engine.setUseClientMode(false);
+                engine.setNeedClientAuth(tlsNeedClientAuth);
+                engine.setNeedClientAuth(tlsWantClientAuth);
+                return engine;
+            }
+
+            private TrustManager[] initTrustStore()
+                    throws KeyStoreException, CertificateException, NoSuchAlgorithmException, IOException {
+                KeyStore trustStore = KeyStore.getInstance("JKS");
+                trustStore.load(null, null);
+                loadCertificates(trustStore, tlsClientAuthCertFile, CertificateFactory.getInstance("X.509"));
+                LOG.info("TrustStore: " + trustStore + " aliases: " + join(trustStore.aliases()));
+                TrustManagerFactory instance = TrustManagerFactory
+                        .getInstance(TrustManagerFactory.getDefaultAlgorithm());
+                instance.init(trustStore);
+                return instance.getTrustManagers();
+            }
+
+            private void loadCertificates(KeyStore trustStore, File file, CertificateFactory cf)
+                    throws CertificateException, KeyStoreException, IOException {
+                if (file.isFile()) {
+                    List<Certificate> certChain = Lists
+                            .newArrayList(cf.generateCertificates(new FileInputStream(file)));
+                    for (int i = 0; i < certChain.size(); i++) {
+                        Certificate cert = certChain.get(i);
+                        trustStore.setCertificateEntry(file.getAbsolutePath() + "_" + i, cert);
+                        LOG.debug("adding certificate to truststore:", cert.toString());
+                    }
+                }
+                else if (file.isDirectory()) {
+                    for (Path f : Files.newDirectoryStream(file.toPath())) {
+                        loadCertificates(trustStore, f.toFile(), cf);
+                    }
+
+                }
+            }
+
+            private String join(Enumeration<String> aliases) {
```

Comment:
Please use Guava's [`Joiner`](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Joiner.html) instead of implementing this yourself.